### ISSUE TITLE
ssrnat: add addnCB, addnBC lemmas?

### ISSUE THREAD

**Issue by anton-trunov** [math-comp/math-comp#212]
I needed one of the following lemmas, but it looks like `ssrnat` does not have it. 

```coq
From mathcomp Require Import ssreflect ssrfun ssrnat.
Set Implicit Arguments.
Unset Strict Implicit.
Unset Printing Implicit Defensive.

Lemma addnCB m n : m + (n - m) = m - n + n. 
Proof. by rewrite -maxnE maxnC maxnE addnC. Qed.

Lemma addnBC m n : n - m + m = m - n + n.
Proof. by rewrite addnC addnCB. Qed.
```

Does it make sense to add those to mathcomp? On one hand it's not that hard to prove those on the fly and the lemmas are not that general, and on the other hand, it might be worth to include those for the sake of completeness.

As usual, naming is hard and I tried to choose a name guided by this alternate formulation for `addnBC`:
```coq
Lemma addnBC' : commutative (fun m n => m - n + n).
```

Feel free to close the issue if you think there is no need to bloat the code base.

**Comment by CohenCyril** [comment]
Hi @anton-trunov, I am not sure either,... I guess that for such rules that can be easily obtained on the fly, a good measure is: how many occurrences of this is there in the codebase?

On the other hand, I think it is okay to have any possible 1 2 and 3 variables `nat` statements (using variables, symbols `+`, `-`, `=` ,`<=`, `<` and at most one `.+1` and `.-1`) in the library. I am actually wondering how close we are from having them all?

**Comment by anton-trunov** [comment]
Quite a bit of lemmas are missing (many of them simply because of permutations of `+` due to commutativity and associativity).

I tried to summarize my findings in the following table. Caveat: I might have forgotten to include something, also I didn't prove the new lemmas, so there might be some mistakes there too.

*Legend*: `___lemma` means it's a variation of `lemma` lemma (usually just because of commutativity of addition); `- lemma` means we rewrite `lemma` backwards.

| Conditions     | Equation                  | Lemma                      |
|-------------- |------------------------- |-------------------------- |
|                | 0 + m = m                 | add0n                      |
|                | 0 - m = 0                 | sub0n                      |
|                | 1 + m = m.+1              | add1n                      |
|                | m + 0 = m                 | addn0                      |
|                | m + 1 = m.+1              | addn1                      |
|                | m - 0 = m                 | subn0                      |
|                | m - 1 = m.-1              | subn1                      |
|                | m - m = 0                 | subnn                      |
|                | m.+1 - m = 1              | subSnn                     |
|                | m.+1.-1 = m               | succnK                     |
| 0 < m          | m.-1.+1 = m               | prednK                     |
| n < m          | (m - n.+1).+1 = m - n     | subnSK                     |
|                | (m.+1 - n).-1 = m - n     | subSKn                     |
|                | m + n = n + m             | addnC                      |
|                | m + n.+1 = (m + n).+1     | addnS                      |
|                | m - n.+1 = (m - n).-1     | subnS                      |
|                | m.+1 + n = (m + n).+1     | addSn                      |
|                | m.+1 + n = m + n.+1       | addSnnS                    |
| n <= m         | m.+1 - n = (m - n).+1     | subSn                      |
|                | m.+1 - n.+1 = m - n       | subSS                      |
|                | (m + n) + p = (m + p) + n | addnAC                     |
|                | (m + n) + p = (n + m) + p |                            |
|                | (m + n) + p = (n + p) + m |                            |
|                | (m + n) + p = (p + m) + n |                            |
|                | (m + n) + p = (p + n) + m |                            |
|                | (m + n) + p = m + (n + p) | - addnA                    |
|                | (m + n) + p = m + (p + n) |                            |
|                | (m + n) + p = n + (m + p) |                            |
|                | (m + n) + p = n + (p + m) |                            |
|                | (m + n) + p = p + (m + n) |                            |
|                | (m + n) + p = p + (n + m) |                            |
|                | (m + n) - m = n           | addKn                      |
|                | (m + n) - n = m           | addnK                      |
| p <= m         | (m + n) - p = (m - p) + n | - addnBAC                  |
| p <= n         | (m + n) - p = (n - p) + m | - \_\_\_<sub>addnBAC</sub> |
| p <= n         | (m + n) - p = m + (n - p) | - addnBA                   |
| n <= p         | (m + n) - p = m - (p - n) | - subnBA                   |
| p <= m         | (m + n) - p = n + (m - p) | - \_\_\_<sub>addnBA</sub>  |
| m <= p         | (m + n) - p = n - (p - m) | - \_\_\_<sub>subnBA</sub>  |
|                | (m - n) + n = (n - m) + m | PROPOSED: addnBC           |
| n <= m         | (m - n) + n = m           | subnK                      |
|                | (m - n) + n = m + (n - m) | PROPOSED: addnCB           |
| n <= m         | (m - n) + p = (m + p) - n | addnBAC                    |
| n <= m         | (m - n) + p = (p + m) - n | \_\_\_<sub>addnBAC</sub>   |
| n <= m, n <= p | (m - n) + p = (p - n) + m |                            |
| n <= m, n <= p | (m - n) + p = m + (p - n) | - addnABC                  |
| n <= m, p <= n | (m - n) + p = m - (n - p) |                            |
| m <= n         | (m - n) + p = p           |                            |
|                | (m - n) - p = (m - p) - n | subnAC                     |
|                | (m - n) - p = m - (n + p) | - subnDA                   |
|                | (m - n) - p = m - (p + n) | - \_\_\_<sub>subnDA</sub>  |
|                | m + (n + p) = (m + n) + p | addnA                      |
|                | m + (n + p) = (m + p) + n |                            |
|                | m + (n + p) = (n + m) + p |                            |
|                | m + (n + p) = (n + p) + m |                            |
|                | m + (n + p) = (p + m) + n |                            |
|                | m + (n + p) = (p + n) + m |                            |
|                | m + (n + p) = m + (p + n) |                            |
|                | m + (n + p) = n + (m + p) | addnCA                     |
|                | m + (n + p) = n + (p + m) |                            |
|                | m + (n + p) = p + (m + n) |                            |
|                | m + (n + p) = p + (n + m) |                            |
| m <= n         | m + (n - m) = n           | subnKC                     |
| p <= n         | m + (n - p) = (m + n) - p | addnBA                     |
| p <= m, p <= n | m + (n - p) = (m - p) + n | addnABC                    |
| p <= n         | m + (n - p) = (n + m) - p | \_\_\_<sub>addnBA</sub>    |
| n <= p         | m + (n - p) = m           |                            |
| p <= m, p <= n | m + (n - p) = n + (m - p) | addnBCA                    |
| n <= m         | m - (m - n) = n           | subKn                      |
|                | m - (n + p) = (m - n) - p | subnDA                     |
|                | m - (n + p) = (m - p) - n | \_\_\_<sub>subnDA</sub>    |
| p <= n         | m - (n - p) = (m + p) - n | subnBA                     |
| p <= n, n <= m | m - (n - p) = (m - n) + p |                            |
| p <= n         | m - (n - p) = (p + m) - n | \_\_\_<sub>subnBA</sub>    |
| n <= p         | m - (n - p) = m           |                            |
| p <= n, n <= m | m - (n - p) = p + (m - n) |                            |

**Comment by CohenCyril** [comment]
@anton-trunov It looks to me that this is almost complete now. The ones exclusively with `+` should not be added, since could be covered using an AC plugin (either my ssrAC (https://github.com/math-comp/math-comp/pull/211) or another e.g. AAC_tactics if we make it work with mathcomp). The ones with vanishing `(n - m)` can be dicharged by combining `(eqP le_nm)` with some arithmetic simplification. And lemma `(m - n) + p = (p - n) + m` is `- addnABC` and `addnC`.

The only really missing one would be `m - (n - p) = (m - n) + p` which would better be named `subnBA`,
and `(m - n) + p = m - (n - p)`  and `m - (n - p) = p + (m - n)` are variants of it...
However, `subnBA` is currently mapped to `m - (n - p) = (m + p) - n` which looks weird to me since there is also a commutativity involved, I would rather have  it named `subnCBA`.

**Comment by amahboubi** [comment]
@CohenCyril , as a rule of thumb, I do not think that it is a good idea to justify a merge/inclusion choice by invoking the possible inclusion of future automation. Which does not mean I want to have them all. The table made by @anton-trunov is great however. I think it would be useful to have it somewhere in the wiki. Even better if the empty slots are filled with the corresponding incantation.

**Comment by amahboubi** [comment]
> However, subnBA is currently mapped to m - (n - p) = (m + p) - n which looks weird to me since there is also a commutativity involved, I would rather have it named subnCBA.

I am not so sure. For instance we have:
`Lemma opprB x y : - (x - y) = y - x.`

in `ssralg`.

**Comment by anton-trunov** [comment]
> I think it would be useful to have it somewhere in the wiki. Even better if the empty slots are filled with the corresponding incantation.

Thanks @amahboubi, and good idea! I'll try to fill it in. It would also be good if someone went over the incantations afterwards and checked if those were indeed minimal.

**Comment by CohenCyril** [comment]
@amahboubi you are right for both.
- Actually the best name for `m - (n - p) = (m - n) + p` is simply `subnA` :smile: 
- Missing `+` modulo AC could be named using the same kind of conventions as in my plugin ssrAC :laughing: (e.g. `m + (n + p) = (n + p) + m` is `addnAC231`)

**Comment by CohenCyril** [comment]
Does this need more discussion, or is it just a matter of implementation now?

**Comment by anton-trunov** [comment]
> Does this need more discussion, or is it just a matter of implementation now?

@CohenCyril Sorry, I completely forgot about this issue. I'll resume working on it shortly.

**Comment by CohenCyril** [comment]
Hi @anton-trunov! Sorry for my reactivity, did you plan to prepare a pull request for this issue?

**Comment by CohenCyril** [comment]
Hi @anton-trunov, do you still plan to work on this?

**Comment by anton-trunov** [comment]
Hi @CohenCyril! Sorry for my belated reply. I tried to address this issue in PR #506.

**Comment by chdoc** [comment]
@CohenCyril Should this issue be closed now that #506 has been merged, or is there something else that needs to be done? (That is, is it just a matter of this issue not having been linked to the PR)

**Comment by CohenCyril** [comment]
@chdoc that's a question that I keep asking myself, thanks for speaking it out.
One the one hand, the issue from original post has been addressed, on the other hand [a subsequent post by @anton-trunov](https://github.com/math-comp/math-comp/issues/212#issuecomment-409658821) brings more context to the issue, and deserve discussions and to be addressed.
Hence, let us close this issue and open a new one with the post from @anton-trunov.
